database.py

```diff
@@ -0,0 +1,135 @@
+
+    def execute_with_retry(self, operations, *args, **kwds):
+        max_retries = 5
+        for retry in range(max_retries):
+            try:
+                with self.conn.cursor() as cursor:
+                    return operations(cursor, *args, **kwds)
+            except MySQLdb.OperationalError:
+                if retry == max_retries - 1:
+                    raise
+                else:
+                    self._sleep(2 ** retry)
+                    self._do_connect()
+            else:
+                break
+
+    def execute_with_retry_s(self, sql, *args):
+        def operations(cursor, sql, *args):
+            cursor.execute(sql, args)
+            if cursor.rowcount > 0:
+                return cursor.fetchall()
+            return None
+        return self.execute_with_retry(operations, sql, *args)
+
+    def __getattr__(self, name):
+        return getattr(self.conn, name)
+
+@contextlib.contextmanager
+def ignore_warnings():
+    warnings.filterwarnings('ignore', category = MySQLdb.Warning)
+    yield
+    warnings.resetwarnings()
+
+def _connect(**kwds):
+    return MySQLdb.connect(charset = 'utf8mb4', autocommit = True, **kwds) 
+
+def _connect_to_cd_mysql():
+    kwds = {'read_default_file': REPLICA_MY_CNF}
+    if utils.running_in_tools_labs():
+        kwds['host'] = TOOLS_LABS_CH_MYSQL_HOST
+    return _connect(**kwds)
+
+def _connect_to_wp_mysql(cfg):
+    kwds = {'read_default_file': REPLICA_MY_CNF}
+    if utils.running_in_tools_labs():
+        # Get the project database name (and ultimately the database server's
+        # hostname) from the name of the database we want, as per:
+        # https://wikitech.wikimedia.org/wiki/Help:Tool_Labs/Database#Naming_conventions
+        xxwiki = cfg.database.replace('_p', '')
+        kwds['host'] = '%s.analytics.db.svc.eqiad.wmflabs' % xxwiki
+    return _connect(**kwds)
+
+def _make_tools_labs_dbname(cursor, database, lang_code):
+    cursor.execute("SELECT SUBSTRING_INDEX(USER(), '@', 1)")
+    user = cursor.fetchone()[0]
+    return '%s__%s_%s' % (user, database, lang_code)
+
+def _use(cursor, database, lang_code):
+    cursor.execute('USE %s' % _make_tools_labs_dbname(
+        cursor, database, lang_code))
+
+def init_scratch_db():
+    cfg = config.get_localized_config()
+    def connect_and_initialize():
+        db = _connect_to_cd_mysql()
+        _use(db.cursor(), 'scratch', cfg.lang_code)
+        return db
+    return _RetryingConnection(connect_and_initialize) 
+
+def init_wp_replica_db(lang_code):
+    cfg = config.get_localized_config(lang_code)
+    def connect_and_initialize():
+        db = _connect_to_wp_mysql(cfg)
+        with db.cursor() as cursor:
+            cursor.execute('USE ' + cfg.database)
+        return db
+    return _RetryingConnection(connect_and_initialize)
+
+# Methods for use in batch scripts, not the serving frontend. These set up the
+# databases, help populate the scratch database and swap it with the serving
+# database.
```

Comment:
Rephrase or remove this comment? We don't have a "serving frontend" so it's misleading.

Reply:
Ok!